sdk/cosmosdb/azure_data_cosmos/src/clients/cosmos_client.rs

```diff
@@ -0,0 +1,116 @@
+pub trait CosmosClientMethods {
+    /// Gets a [`DatabaseClient`] that can be used to access the database with the specified ID.
+    fn database(&self, name: impl Into<String>) -> DatabaseClient;
+}
+
+impl CosmosClient {
+    /// Creates a new CosmosClient, using Entra ID authentication.
+    ///
+    /// # Arguments
+    ///
+    /// * `endpoint` - The full URL of the Cosmos DB account, for example `https://myaccount.documents.azure.com/`.
+    /// * `credential` - An implementation of [`TokenCredential`](azure_core::auth::TokenCredential) that can provide an Entra ID token to use when authenticating.
+    /// * `options` - Optional configuration for the client.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use azure_data_cosmos::CosmosClient;
+    ///
+    /// let credential = azure_identity::create_default_credential().unwrap();
+    /// let client = CosmosClient::new("https://myaccount.documents.azure.com/", credential, None).unwrap();
+    /// ```
+    pub fn new(
+        endpoint: impl AsRef<str>,
+        credential: Arc<dyn TokenCredential>,
+        options: Option<CosmosClientOptions>,
+    ) -> azure_core::Result<Self> {
+        let options = options.unwrap_or_default();
+        Ok(Self {
+            endpoint: endpoint.as_ref().parse()?,
+            pipeline: create_pipeline(
+                AuthorizationPolicy::from_token_credential(credential),
+                options.client_options.clone(),
+            ),
+            options,
+        })
+    }
+
+    /// Creates a new CosmosClient, using key authentication.
+    ///
+    /// # Arguments
+    ///
+    /// * `endpoint` - The full URL of the Cosmos DB account, for example `https://myaccount.documents.azure.com/`.
+    /// * `key` - The key to use when authenticating.
+    /// * `options` - Optional configuration for the client.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// let client = CosmosClient::with_shared_key("https://myaccount.documents.azure.com/", "my_key", None)?;
+    /// ```
+    #[cfg(feature = "key-auth")]
+    pub fn with_key(
+        endpoint: impl AsRef<str>,
+        key: impl Into<Secret>,
+        options: Option<CosmosClientOptions>,
+    ) -> azure_core::Result<Self> {
+        let options = options.unwrap_or_default();
+        Ok(Self {
+            endpoint: endpoint.as_ref().parse()?,
+            pipeline: create_pipeline(
+                AuthorizationPolicy::from_shared_key(key.into()),
+                options.client_options.clone(),
+            ),
+            options,
+        })
+    }
+
+    /// Gets the endpoint of the database account this client is connected to.
+    pub fn endpoint(&self) -> &Url {
+        &self.endpoint
+    }
+}
+
+impl CosmosClientMethods for CosmosClient {
+    /// Gets a [`DatabaseClient`] that can be used to access the database with the specified ID.
+    fn database(&self, id: impl Into<String>) -> DatabaseClient {
```

Comment:
[Current guideline](https://azure.github.io/azure-sdk/rust_introduction.html#rust-subclients-suffix) states this should be named `database_client`. We can discuss, but this was done to avoid confusion about what it returns since you rarely declare a variable type in Rust.